### ISSUE TITLE
regression: yarn fails to install the right version in node_modules/.bin

### ISSUE THREAD

**Issue by evanw** [yarnpkg/yarn#3837]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

If package A requires B and C, and B requires an older version of C, then the older version will be linked into A's node_modules/.bin instead of the newer one.

**If the current behavior is a bug, please provide the steps to reproduce.**

This test case is reduced from a real issue that we're having on our CI servers (we recently switched from npm to yarn).

```sh
$ cat > package.json
{
  "devDependencies": {
    "rollup-plugin-typescript": "^0.8.1",
    "typescript": "^2.4.1"
  }
}
```

```sh
$ brew switch yarn 0.27.5
Cleaning /usr/local/Cellar/yarn/0.24.6
Cleaning /usr/local/Cellar/yarn/0.27.5
2 links created for /usr/local/Cellar/yarn/0.27.5
```

```sh
$ yarn --version
0.27.5
```

```sh
$ rm -fr node_modules/ yarn.lock
```

```sh
$ yarn
yarn install v0.27.5
warning package.json: No license field
info No lockfile found.
warning No license field
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Saved lockfile.
Done in 1.29s.
```

```sh
$ yarn run tsc -- --version
yarn run v0.27.5
warning package.json: No license field
$ "/Users/evan/test/node_modules/.bin/tsc" "--version"
Version 1.8.10
Done in 0.28s.
```

Problem: the top-level package needs `tsc --version` to return `2.4.1` not `1.8.10`. This is the problem:

```sh
$ ls -l node_modules/.bin
total 16
lrwxr-xr-x  1 evan  staff  59 Jul  5 20:56 tsc -> ../rollup-plugin-typescript/node_modules/typescript/bin/tsc
lrwxr-xr-x  1 evan  staff  64 Jul  5 20:56 tsserver -> ../rollup-plugin-typescript/node_modules/typescript/bin/tsserver
```

**What is the expected behavior?**

This appears to be a regression in yarn. I'll compare this with an older version of yarn below, which seems to have the right behavior (the behavior that matches npm).

```sh
$ brew switch yarn 0.24.6
Cleaning /usr/local/Cellar/yarn/0.24.6
Cleaning /usr/local/Cellar/yarn/0.27.5
2 links created for /usr/local/Cellar/yarn/0.24.6
```

```sh
$ yarn --version
0.24.6
```

```sh
$ rm -fr node_modules/ yarn.lock
```

```sh
$ yarn
yarn install v0.24.6
info No lockfile found.
warning No license field
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 📃  Building fresh packages...
success Saved lockfile.
✨  Done in 1.03s.
```

```sh
$ yarn run tsc -- --version
yarn run v0.24.6
$ "/Users/evan/test/node_modules/.bin/tsc" --version
Version 2.4.1
✨  Done in 0.34s.
```

With this older version, the top-level package does get `tsc --version` to return `2.4.1`. The bin folder looks good this time:

```sh
$ ls -l node_modules/.bin
total 16
lrwxr-xr-x  1 evan  staff  21 Jul  5 21:02 tsc -> ../typescript/bin/tsc
lrwxr-xr-x  1 evan  staff  26 Jul  5 21:02 tsserver -> ../typescript/bin/tsserver
```

**Please mention your node.js, yarn and operating system version.**

```sh
$ node --version
v4.8.2
$ npm --version
5.0.3
$ yarn --version
0.24.6
$ sw_vers
ProductName:  Mac OS X
ProductVersion: 10.11.6
BuildVersion: 15G1510
```

**Comment by mblair** [comment]
I think this is the same as #3801.

**Comment by SpainTrain** [comment]
> I think this is the same as #3801.

Yup, appears to be fixed in v0.28.1 (https://github.com/yarnpkg/yarn/releases/tag/v0.28.1)

**Comment by gkalpak** [comment]
This should be fixed by #3877.
